Flash API: new information getter

de-nordic opened the pull request:
First commit in this PR proposes change to flash API that introduces API calls for obtaining information about flash, like size, page count and page information.

This change, contrary to `flash_api_pages_layout`, is non optional addition and does not allow directly returning objects from drivers, instead allows to query the driver to fill in information.

The addition of API calls:
```
ssize_t flash_get_page_count(const struct device *dev);
ssize_t flash_get_size(const struct device *dev);
int flash_get_page_info(const struct device *dev, off_t offset, struct flash_page_info *fpi);
```
allows access to information on flash size and page count, and information on single page.
The commit changes also the `flash_parameters` structure with addition of `flags` and `max_page_size`;
there is one flag defined so far: `FPF_NON_UNIFORM_LAYOUT` that is set for devices with non-uniform layout and should allow code using the flash device to decide how to work with the device.

The second commit is test of changes.

This is proposed implementation of: #36485

**Rationale for changes**
Original idea has been to implement  "virtual" flash device, but the idea has been dropped. Nevertheless the API change allows more efficient checking for device layout.
~~The main goal of the change is to simplify obtaining information on flash devices, and to be able to define "virtual" devices that could be placed over the other flash devices to perform some actions, which existence of `flash_api_pages_layout` does allow, as the "virtual" device is not able to easily override values returned by the call.
This change should allow to, for example, define partition or encryption device, that will use the same flash API calls, which means that any function that is able to perform actions on a device of flash type, will be usable with the "virtual" device, without knowing that it is not working on a real device.~~

~~Example of the "virtual" device that defines partitions over real device: https://github.com/zephyrproject-rtos/zephyr/pull/36487~~

TODO:
- [x] test cases;
- [x] apply to all drivers.

de-nordic committed:
drivers/flash: Flash API: device information getters

The commit adds flowing flash_get_* API functions:
  flash_get_size -- get complete size of flash in bytes;
  flash_get_page_count -- get number of flash pages on devices;
  flash_get_page_info -- get flash page info by offset

The new structure type flash_page_info has been introduced, that
returns information on single page of flash.

It also extends flash_parameters structure by addition of:
  flags -- FPF flags that characterize device.

Following flags has been added:
  FPF_NON_UNIFORM_PAGE_LAYOUT -- device has varying page size
    size need to set;
  FPF_GAPS_IN_LAYOUT -- device addressing has gaps.

This API changes are supposed to replace current flash_pages* API
calls and allow define virtual flash devices, that will be able
to use flash API but at the same time hide from users that they
area for example just limiting access or providing encryption.

In case of uniform layout devices, the API should allow quicker
access to information.

The replacement for flash_get_page_info_by_idx and flash_page_foreach
may be easily implemented using the new API.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/nrf: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_simulator:  Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/drivers/flash_api: Flash API tests

The commit adds tests for Flash API; the tests are preformed
on mocked device to check whether callbacks are properly called
and expected values are returned.
Currently the flash_get_page_info, flash_get_page_count, and
flash_get_size are tested.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu commented on include/drivers/flash.h:
```diff
@@ -489,6 +545,74 @@ static inline const struct flash_parameters *z_impl_flash_get_parameters(const s
+ *	   -EINVAL if offset is outside of flash device;
+ *	   other negative errno code on error.
+ */
+__syscall int flash_get_page_info(const struct device *dev, off_t off, struct flash_page_info *fpi);
```
copy from: https://github.com/zephyrproject-rtos/zephyr/pull/36487#discussion_r664564531
`nvlsianpu:`
isn't this similar to flash_get_page_info_by_offs()
`de-nordic:`
I have added the `flash_get_page_info`, as a part of https://github.com/zephyrproject-rtos/zephyr/issues/36485, so that it would be possible to override results returned by such call, which is not possible with the current flash_api_get_layout, because it gives direct access to objects defined within driver.
I have not removed the original calls that do layout checks, because this is proposal and I did not want to remove them yet.
The `flash_get_page_info_by_offs` and the `flash_get_page_info_by_idx`, and the `foreachs`, in case of the  https://github.com/zephyrproject-rtos/zephyr/issues/36485, would be reimplemented as non-system util library, and the CONFIG that enables them would be still available.
All these functions would be reimplemented around the new `flash_get_page_info`, `flash_get_page_count` and `flash_get_size`.
`nvlsianpu:`
Ok, understood

nvlsianpu replied:
This is all fine to me.

nvlsianpu commented on include/drivers/flash.h:
```diff
@@ -133,6 +196,9 @@ __subsystem struct flash_driver_api {
 	flash_api_erase erase;
 	flash_api_write_protection write_protection;
 	flash_api_get_parameters get_parameters;
+	flash_api_get_page_info get_page_info;
```
copy from: https://github.com/zephyrproject-rtos/zephyr/pull/36487#discussion_r664608864
`nvlsianpu:`
this is similar to flash-layout-get implementation api. Shouldn't it be optional?

`de-nordic:`
I was considering adding this as constant addition as the info is needed anyway, for example you can not do erase without knowing the page info, also this can be used to get alignment as it always returns info for aligned page that contains the offset.
In case of most drivers, the implementation will be quite efficient, because drivers that know about uniform layout will just do something like this:
```
if (offset < TOTAL_FLASH_SIZE) {
   fpi->offset = offset & ~(CONSTANT_PAGE_SIZE -1);
   fpi->size = CONSTANT_PAGE_SIZE);
}
```
On the other side the non uniform layout is also quick to obtain:
```
if (offset < HUGE_PAGE_OFFSET) {
   fpi->offset = offset & ~(CONSTANT_PAGE_SIZE -1);
   fpi->size = CONSTANT_PAGE_SIZE);
} else (offset < TOTAL_FLASH_SIZE) {
   fpi->offset = HUGE_PAGE_OFFSET;
   fpi->size = HUGE_PAGE_SIZE;
}
```
Because driver exactly knows the layout, in most cases, it does not have to iterate through all pages to get to the offset that is interesting for it.

nvlsianpu replied:
I think you are right. Option was introduced when not all devices implements the flash-get-layout API. Nowday it is almost mandatory.

de-nordic committed:
drivers/flash/layout: Reimplement layout functions with new API

The addition of Flash API calls get_page_info, get_page_count
and get_size, visible as flash_get_page_info, flash_get_page_count
and flash_get_size respectively, allows to implement
flash_get_page_info_by_idx, flash_get_page_info_by_offs and
flash_page_foreach without enabling the page_layout callback.
The commit provides implementation of the functions.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/nrf_qspi_nor:  Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/esp32: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/at45: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/stm32_qspi: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/gecko: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flexspi_nor: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flexspi_x25um51345g: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_sam: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/soc_flash_lpc: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/soc_flash_mcux: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/soc_flash_rv32m1: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/soc_flash_nios2_qspi: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/spi_nor: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32wbx: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32l4x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32g4x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32g0x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32f4x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32_v1: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32h7x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32f7x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32f2x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_stm32l5x: Add flash get_page_info API call

The commit adds implementation of get_page_info, get_page_count
and get_size to the flash driver.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
@PerMac Can you help me with the CI, I have no idea what is it going with the parsing error.

PerMac commented:
I see an `error ERROR   - tests/drivers/flash_api/drivers.flash.api.build_only - platform_allow - unrecognized platform - \`
" unrecognized platform" is a recent check added by @gopiotr to verify if a platform name which is used in sever possible contexts actually exists in zephyr. Here it point to platform_allow list in  tests/ests/driverdrivers/flash_api/ yaml It seems it doesn't like that you're using \ to continue a line as twister thinks "\" is a platform. BTW maybe you can think of same other means of filtering platforms than "platform_allow"? such list are hard to maintain and block anything what's not on it, e.g any future platforms.

de-nordic commented:
> BTW maybe you can think of same other means of filtering platforms than "platform_allow"? such list are hard to maintain and block anything what's not on it, e.g any future platforms.

All I want to do is make sure the thing gets at least build for these platforms, is there easier way?

PerMac commented:
> All I want to do is make sure the thing gets at least build for these platforms, is there easier way?

Then you want to put these platforms into `integration_platforms` and not in the `platform_allow`.

de-nordic commented:
@erwango @nvlsianpu Can you take a look again? The PR went through dev-review twice and got accepted but lacks reviews from platform specialists.
